QoB is raising "Stream is already closed" when closing an FS stream

### What happened?

https://ci.azure.hail.is/batches/3775837/jobs/45

```
E           java.util.concurrent.ExecutionException: java.lang.RuntimeException: Stream is already closed.
E           	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
E           	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
E           	at is.hail.backend.service.ServiceBackend.parallelizeAndComputeWithIndex(ServiceBackend.scala:150)
E           	at is.hail.backend.BackendUtils.collectDArray(BackendUtils.scala:44)
E           	at __C256669Compiled.__m256730split_CollectDistributedArray(Emit.scala)
E           	at __C256669Compiled.__m256689split_Let(Emit.scala)
E           	at __C256669Compiled.apply(Emit.scala)
E           	at is.hail.expr.ir.CompileAndEvaluate$.$anonfun$_apply$7(CompileAndEvaluate.scala:74)
E           	at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.CompileAndEvaluate$._apply(CompileAndEvaluate.scala:74)
E           	at is.hail.expr.ir.CompileAndEvaluate$.$anonfun$apply$1(CompileAndEvaluate.scala:19)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.CompileAndEvaluate$.apply(CompileAndEvaluate.scala:19)
E           	at is.hail.expr.ir.lowering.LowerDistributedSort$.distributedSort(LowerDistributedSort.scala:163)
E           	at is.hail.backend.service.ServiceBackend.lowerDistributedSort(ServiceBackend.scala:354)
E           	at is.hail.backend.Backend.lowerDistributedSort(Backend.scala:100)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShuffles$.$anonfun$apply$1(LowerAndExecuteShuffles.scala:23)
E           	at is.hail.expr.ir.RewriteBottomUp$.$anonfun$apply$4(RewriteBottomUp.scala:26)
E           	at is.hail.utils.StackSafe$More.advance(StackSafe.scala:60)
E           	at is.hail.utils.StackSafe$.run(StackSafe.scala:16)
E           	at is.hail.utils.StackSafe$StackFrame.run(StackSafe.scala:32)
E           	at is.hail.expr.ir.RewriteBottomUp$.apply(RewriteBottomUp.scala:36)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShuffles$.apply(LowerAndExecuteShuffles.scala:20)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShufflesPass.transform(LoweringPass.scala:157)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$3(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$1(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply(LoweringPass.scala:14)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply$(LoweringPass.scala:13)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShufflesPass.apply(LoweringPass.scala:151)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1(LoweringPipeline.scala:22)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1$adapted(LoweringPipeline.scala:20)
E           	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
E           	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
E           	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.apply(LoweringPipeline.scala:20)
E           	at is.hail.expr.ir.lowering.EvalRelationalLets$.execute$1(EvalRelationalLets.scala:10)
E           	at is.hail.expr.ir.lowering.EvalRelationalLets$.lower$1(EvalRelationalLets.scala:18)
E           	at is.hail.expr.ir.lowering.EvalRelationalLets$.apply(EvalRelationalLets.scala:37)
E           	at is.hail.expr.ir.lowering.EvalRelationalLetsPass.transform(LoweringPass.scala:147)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$3(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$1(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply(LoweringPass.scala:14)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply$(LoweringPass.scala:13)
E           	at is.hail.expr.ir.lowering.EvalRelationalLetsPass.apply(LoweringPass.scala:141)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1(LoweringPipeline.scala:22)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1$adapted(LoweringPipeline.scala:20)
E           	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
E           	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
E           	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.apply(LoweringPipeline.scala:20)
E           	at is.hail.backend.service.ServiceBackend.execute(ServiceBackend.scala:310)
E           	at is.hail.backend.service.ServiceBackend.execute(ServiceBackend.scala:346)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$12(ServiceBackend.scala:698)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.withIRFunctionsReadFromInput(ServiceBackend.scala:801)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$11(ServiceBackend.scala:696)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$2(ServiceBackend.scala:654)
E           	at is.hail.backend.ExecuteContext$.$anonfun$scoped$3(ExecuteContext.scala:75)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.ExecuteContext$.$anonfun$scoped$2(ExecuteContext.scala:75)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.annotations.RegionPool$.scoped(RegionPool.scala:17)
E           	at is.hail.backend.ExecuteContext$.scoped(ExecuteContext.scala:63)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$1(ServiceBackend.scala:644)
E           	at is.hail.utils.ExecutionTimer$.time(ExecutionTimer.scala:52)
E           	at is.hail.utils.ExecutionTimer$.logTime(ExecutionTimer.scala:59)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.withExecuteContext$1(ServiceBackend.scala:631)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.executeOneCommand(ServiceBackend.scala:693)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$6(ServiceBackend.scala:459)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$6$adapted(ServiceBackend.scala:458)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$5(ServiceBackend.scala:458)
E           	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E           	at is.hail.services.package$.retryTransientErrors(package.scala:124)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$4(ServiceBackend.scala:458)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$4$adapted(ServiceBackend.scala:456)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$3(ServiceBackend.scala:456)
E           	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E           	at is.hail.services.package$.retryTransientErrors(package.scala:124)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.main(ServiceBackend.scala:456)
E           	at is.hail.backend.service.Main$.main(Main.scala:15)
E           	at is.hail.backend.service.Main.main(Main.scala)
E           	at sun.reflect.GeneratedMethodAccessor90.invoke(Unknown Source)
E           	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E           	at java.lang.reflect.Method.invoke(Method.java:498)
E           	at is.hail.JVMEntryway$1.run(JVMEntryway.java:119)
E           	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
E           	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E           	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
E           	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E           	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E           	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E           	at java.lang.Thread.run(Thread.java:750)
E           
E           java.lang.RuntimeException: Stream is already closed.
E           	at com.azure.storage.common.StorageOutputStream.checkStreamState(StorageOutputStream.java:79)
E           	at com.azure.storage.common.StorageOutputStream.flush(StorageOutputStream.java:89)
E           	at is.hail.io.fs.AzureStorageFS$$anon$3.close(AzureStorageFS.scala:291)
E           	at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
E           	at is.hail.utils.package$.using(package.scala:640)
E           	at is.hail.io.fs.FS.writePDOS(FS.scala:428)
E           	at is.hail.io.fs.FS.writePDOS$(FS.scala:427)
E           	at is.hail.io.fs.RouterFS.writePDOS(RouterFS.scala:3)
E           	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$3(ServiceBackend.scala:114)
E           	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$3$adapted(ServiceBackend.scala:114)
E           	at is.hail.backend.service.ServiceBackend$$anon$2.$anonfun$call$1(ServiceBackend.scala:122)
E           	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E           	at is.hail.services.package$.retryTransientErrors(package.scala:124)
E           	at is.hail.backend.service.ServiceBackend$$anon$2.call(ServiceBackend.scala:122)
E           	at is.hail.backend.service.ServiceBackend$$anon$2.call(ServiceBackend.scala:119)
E           	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E           	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E           	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E           	at java.lang.Thread.run(Thread.java:750)
E           
E           
E           
E           
E           Hail version: 0.2.115-f6017673dbb6
E           Error summary: RuntimeException: Stream is already closed.
```

### Version

0.2.115-f6017673dbb6

### Relevant log output

```shell
________________________________ test_spectra_4 ________________________________
[gw2] linux -- Python 3.8.10 /usr/bin/python3

    def test_spectra_4():
>       spectra_helper(spec4)

test/hail/methods/test_pca.py:229: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/hail/methods/test_pca.py:172: in spectra_helper
    hail_V = (np.array(scores.scores.collect()) / singulars).T
<decorator-gen-538>:2: in collect
    ???
/usr/local/lib/python3.8/dist-packages/hail/typecheck/check.py:584: in wrapper
    return __original_func(*args_, **kwargs_)
/usr/local/lib/python3.8/dist-packages/hail/expr/expressions/base_expression.py:1132: in collect
    return hl.eval(e)
<decorator-gen-702>:2: in eval
    ???
/usr/local/lib/python3.8/dist-packages/hail/typecheck/check.py:584: in wrapper
    return __original_func(*args_, **kwargs_)
/usr/local/lib/python3.8/dist-packages/hail/expr/expressions/expression_utils.py:191: in eval
    return eval_timed(expression)[0]
<decorator-gen-700>:2: in eval_timed
    ???
/usr/local/lib/python3.8/dist-packages/hail/typecheck/check.py:584: in wrapper
    return __original_func(*args_, **kwargs_)
/usr/local/lib/python3.8/dist-packages/hail/expr/expressions/expression_utils.py:161: in eval_timed
    return Env.backend().execute(MakeTuple([ir]), timed=True)[0]
/usr/local/lib/python3.8/dist-packages/hail/backend/service_backend.py:490: in execute
    return self._cancel_on_ctrl_c(self._async_execute(ir, timed=timed))
/usr/local/lib/python3.8/dist-packages/hail/backend/service_backend.py:481: in _cancel_on_ctrl_c
    return async_to_blocking(coro)
/usr/local/lib/python3.8/dist-packages/hailtop/utils/utils.py:152: in async_to_blocking
    return loop.run_until_complete(task)
/usr/local/lib/python3.8/dist-packages/nest_asyncio.py:90: in run_until_complete
    return f.result()
/usr/lib/python3.8/asyncio/futures.py:178: in result
    raise self._exception
/usr/lib/python3.8/asyncio/tasks.py:280: in __step
    result = coro.send(None)
/usr/local/lib/python3.8/dist-packages/hail/backend/service_backend.py:509: in _async_execute
    _, resp, timings = await self._rpc('execute(...)', inputs, ir=ir, progress=progress)
/usr/local/lib/python3.8/dist-packages/hail/backend/service_backend.py:451: in _rpc
    result_bytes = await retry_transient_errors(self._read_output, ir, iodir + '/out')
/usr/local/lib/python3.8/dist-packages/hailtop/utils/utils.py:781: in retry_transient_errors
    return await retry_transient_errors_with_debug_string('', 0, f, *args, **kwargs)
/usr/local/lib/python3.8/dist-packages/hailtop/utils/utils.py:794: in retry_transient_errors_with_debug_string
    return await f(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <hail.backend.service_backend.ServiceBackend object at 0x7ff6aa3b15b0>
ir = <hail.ir.ir.MakeTuple object at 0x7ff6a73b8d30>
output_uri = 'hail-az://haildevtest/test/tmp/hail/mGZ0y8JSUQM6CVs20XzdIu/eLwQUTyLRv/out'

    async def _read_output(self, ir: Optional[BaseIR], output_uri: str) -> bytes:
        assert self._batch
    
        try:
            driver_output = await self._async_fs.open(output_uri)
        except FileNotFoundError as exc:
            raise FatalError('Hail internal error. Please contact the Hail team and provide the following information.\n\n' + yamlx.dump({
                'service_backend_debug_info': self.debug_info(),
                'batch_debug_info': await self._batch.debug_info()
            })) from exc
    
        async with driver_output as outfile:
            success = await read_bool(outfile)
            if success:
                return await read_bytes(outfile)
    
            short_message = await read_str(outfile)
            expanded_message = await read_str(outfile)
            error_id = await read_int(outfile)
    
            reconstructed_error = fatal_error_from_java_error_triplet(short_message, expanded_message, error_id)
            if ir is None:
                raise reconstructed_error
>           raise reconstructed_error.maybe_user_error(ir)
E           hail.utils.java.FatalError: RuntimeException: Stream is already closed.
E           
E           Java stack trace:
E           java.util.concurrent.ExecutionException: java.lang.RuntimeException: Stream is already closed.
E           	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
E           	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
E           	at is.hail.backend.service.ServiceBackend.parallelizeAndComputeWithIndex(ServiceBackend.scala:150)
E           	at is.hail.backend.BackendUtils.collectDArray(BackendUtils.scala:44)
E           	at __C256669Compiled.__m256730split_CollectDistributedArray(Emit.scala)
E           	at __C256669Compiled.__m256689split_Let(Emit.scala)
E           	at __C256669Compiled.apply(Emit.scala)
E           	at is.hail.expr.ir.CompileAndEvaluate$.$anonfun$_apply$7(CompileAndEvaluate.scala:74)
E           	at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.CompileAndEvaluate$._apply(CompileAndEvaluate.scala:74)
E           	at is.hail.expr.ir.CompileAndEvaluate$.$anonfun$apply$1(CompileAndEvaluate.scala:19)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.CompileAndEvaluate$.apply(CompileAndEvaluate.scala:19)
E           	at is.hail.expr.ir.lowering.LowerDistributedSort$.distributedSort(LowerDistributedSort.scala:163)
E           	at is.hail.backend.service.ServiceBackend.lowerDistributedSort(ServiceBackend.scala:354)
E           	at is.hail.backend.Backend.lowerDistributedSort(Backend.scala:100)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShuffles$.$anonfun$apply$1(LowerAndExecuteShuffles.scala:23)
E           	at is.hail.expr.ir.RewriteBottomUp$.$anonfun$apply$4(RewriteBottomUp.scala:26)
E           	at is.hail.utils.StackSafe$More.advance(StackSafe.scala:60)
E           	at is.hail.utils.StackSafe$.run(StackSafe.scala:16)
E           	at is.hail.utils.StackSafe$StackFrame.run(StackSafe.scala:32)
E           	at is.hail.expr.ir.RewriteBottomUp$.apply(RewriteBottomUp.scala:36)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShuffles$.apply(LowerAndExecuteShuffles.scala:20)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShufflesPass.transform(LoweringPass.scala:157)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$3(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$1(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply(LoweringPass.scala:14)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply$(LoweringPass.scala:13)
E           	at is.hail.expr.ir.lowering.LowerAndExecuteShufflesPass.apply(LoweringPass.scala:151)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1(LoweringPipeline.scala:22)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1$adapted(LoweringPipeline.scala:20)
E           	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
E           	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
E           	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.apply(LoweringPipeline.scala:20)
E           	at is.hail.expr.ir.lowering.EvalRelationalLets$.execute$1(EvalRelationalLets.scala:10)
E           	at is.hail.expr.ir.lowering.EvalRelationalLets$.lower$1(EvalRelationalLets.scala:18)
E           	at is.hail.expr.ir.lowering.EvalRelationalLets$.apply(EvalRelationalLets.scala:37)
E           	at is.hail.expr.ir.lowering.EvalRelationalLetsPass.transform(LoweringPass.scala:147)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$3(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.$anonfun$apply$1(LoweringPass.scala:16)
E           	at is.hail.utils.ExecutionTimer.time(ExecutionTimer.scala:81)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply(LoweringPass.scala:14)
E           	at is.hail.expr.ir.lowering.LoweringPass.apply$(LoweringPass.scala:13)
E           	at is.hail.expr.ir.lowering.EvalRelationalLetsPass.apply(LoweringPass.scala:141)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1(LoweringPipeline.scala:22)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.$anonfun$apply$1$adapted(LoweringPipeline.scala:20)
E           	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
E           	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
E           	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
E           	at is.hail.expr.ir.lowering.LoweringPipeline.apply(LoweringPipeline.scala:20)
E           	at is.hail.backend.service.ServiceBackend.execute(ServiceBackend.scala:310)
E           	at is.hail.backend.service.ServiceBackend.execute(ServiceBackend.scala:346)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$12(ServiceBackend.scala:698)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.withIRFunctionsReadFromInput(ServiceBackend.scala:801)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$11(ServiceBackend.scala:696)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$2(ServiceBackend.scala:654)
E           	at is.hail.backend.ExecuteContext$.$anonfun$scoped$3(ExecuteContext.scala:75)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.ExecuteContext$.$anonfun$scoped$2(ExecuteContext.scala:75)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.annotations.RegionPool$.scoped(RegionPool.scala:17)
E           	at is.hail.backend.ExecuteContext$.scoped(ExecuteContext.scala:63)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.$anonfun$executeOneCommand$1(ServiceBackend.scala:644)
E           	at is.hail.utils.ExecutionTimer$.time(ExecutionTimer.scala:52)
E           	at is.hail.utils.ExecutionTimer$.logTime(ExecutionTimer.scala:59)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.withExecuteContext$1(ServiceBackend.scala:631)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2.executeOneCommand(ServiceBackend.scala:693)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$6(ServiceBackend.scala:459)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$6$adapted(ServiceBackend.scala:458)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$5(ServiceBackend.scala:458)
E           	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E           	at is.hail.services.package$.retryTransientErrors(package.scala:124)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$4(ServiceBackend.scala:458)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$4$adapted(ServiceBackend.scala:456)
E           	at is.hail.utils.package$.using(package.scala:635)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.$anonfun$main$3(ServiceBackend.scala:456)
E           	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E           	at is.hail.services.package$.retryTransientErrors(package.scala:124)
E           	at is.hail.backend.service.ServiceBackendSocketAPI2$.main(ServiceBackend.scala:456)
E           	at is.hail.backend.service.Main$.main(Main.scala:15)
E           	at is.hail.backend.service.Main.main(Main.scala)
E           	at sun.reflect.GeneratedMethodAccessor90.invoke(Unknown Source)
E           	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
E           	at java.lang.reflect.Method.invoke(Method.java:498)
E           	at is.hail.JVMEntryway$1.run(JVMEntryway.java:119)
E           	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
E           	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E           	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
E           	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E           	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E           	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E           	at java.lang.Thread.run(Thread.java:750)
E           
E           java.lang.RuntimeException: Stream is already closed.
E           	at com.azure.storage.common.StorageOutputStream.checkStreamState(StorageOutputStream.java:79)
E           	at com.azure.storage.common.StorageOutputStream.flush(StorageOutputStream.java:89)
E           	at is.hail.io.fs.AzureStorageFS$$anon$3.close(AzureStorageFS.scala:291)
E           	at java.io.FilterOutputStream.close(FilterOutputStream.java:159)
E           	at is.hail.utils.package$.using(package.scala:640)
E           	at is.hail.io.fs.FS.writePDOS(FS.scala:428)
E           	at is.hail.io.fs.FS.writePDOS$(FS.scala:427)
E           	at is.hail.io.fs.RouterFS.writePDOS(RouterFS.scala:3)
E           	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$3(ServiceBackend.scala:114)
E           	at is.hail.backend.service.ServiceBackend.$anonfun$parallelizeAndComputeWithIndex$3$adapted(ServiceBackend.scala:114)
E           	at is.hail.backend.service.ServiceBackend$$anon$2.$anonfun$call$1(ServiceBackend.scala:122)
E           	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
E           	at is.hail.services.package$.retryTransientErrors(package.scala:124)
E           	at is.hail.backend.service.ServiceBackend$$anon$2.call(ServiceBackend.scala:122)
E           	at is.hail.backend.service.ServiceBackend$$anon$2.call(ServiceBackend.scala:119)
E           	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E           	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
E           	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
E           	at java.lang.Thread.run(Thread.java:750)
E           
E           
E           
E           
E           Hail version: 0.2.115-f6017673dbb6
E           Error summary: RuntimeException: Stream is already closed.

/usr/local/lib/python3.8/dist-packages/hail/backend/service_backend.py:477: FatalError
------------------------------ Captured log call -------------------------------
INFO     batch_client.aioclient:aioclient.py:753 created batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 1/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 2/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 3/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 4/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Iterations complete. Computing local QR
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 _reduced_svd: Computing local SVD
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 blanczos_pca: SVD Complete. Computing conversion to PCs.
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 1/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 2/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 3/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Beginning iteration 4/4
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 krylov_factorization: Iterations complete. Computing local QR
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 _reduced_svd: Computing local SVD
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     backend.service_backend:java.py:190 blanczos_pca: SVD Complete. Computing conversion to PCs.
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
INFO     batch_client.aioclient:aioclient.py:770 updated batch 3776913
```

--- a/hail/build.gradle
+++ b/hail/build.gradle
@@ -199,30 +199,14 @@ dependencies {
     }
 
     bundled 'commons-io:commons-io:2.11.0'
-    bundled(group: 'com.azure', name: 'azure-storage-blob', version: '12.13.0') {
-        exclude group: 'com.fasterxml.jackson'
-        exclude group: 'com.fasterxml.jackson.core'
-        exclude group: 'com.fasterxml.jackson.databind'
-    }
-
-    bundled(group: 'com.azure', name: 'azure-core-http-netty', version: '1.10.0') {
-        exclude group: 'com.fasterxml.jackson'
-        exclude group: 'com.fasterxml.jackson.core'
-        exclude group: 'com.fasterxml.jackson.databind'
-    }
-
-
-    bundled(group: 'com.azure', name: 'azure-identity', version:'1.2.1') {
-        exclude group: 'com.fasterxml.jackson'
-        exclude group: 'com.fasterxml.jackson.core'
-        exclude group: 'com.fasterxml.jackson.databind'
-    }
 
     bundled group: 'org.freemarker', name: 'freemarker', version: '2.3.31'
 
     bundled 'com.kohlschutter.junixsocket:junixsocket-core:2.6.1'
 
     bundled 'com.github.luben:zstd-jni:1.5.2-1'
+
+    bundled project(path: ':shadedazure', configuration: 'shadow')
 }
 
 task(checkSettings) doLast {

--- a/hail/scripts/upload_qob_jar.sh
+++ b/hail/scripts/upload_qob_jar.sh
@@ -22,5 +22,5 @@ else
     JAR_LOCATION="${TEST_STORAGE_URI}/${NAMESPACE}/jars/${TOKEN}/${REVISION}.jar"
 fi
 
-az storage blob upload --file ${SHADOW_JAR} --account-name haildevbatch --container-name query --name jars/dking/lfrxrx935eaq/74f44922afbdceb900661e28011ed4004e6ecb25.jar
+python3 -m hailtop.aiotools.copy -vvv 'null' '[{"from":"'${SHADOW_JAR}'", "to":"'${JAR_LOCATION}'"}]'
 echo ${JAR_LOCATION} > ${PATH_FILE}

--- a/hail/scripts/upload_qob_jar.sh
+++ b/hail/scripts/upload_qob_jar.sh
@@ -22,5 +22,5 @@ else
     JAR_LOCATION="${TEST_STORAGE_URI}/${NAMESPACE}/jars/${TOKEN}/${REVISION}.jar"
 fi
 
-gcloud storage cp ${SHADOW_JAR} ${JAR_LOCATION}
+az storage blob upload --file ${SHADOW_JAR} --account-name haildevbatch --container-name query --name jars/dking/lfrxrx935eaq/74f44922afbdceb900661e28011ed4004e6ecb25.jar
 echo ${JAR_LOCATION} > ${PATH_FILE}

--- a/hail/settings.gradle
+++ b/hail/settings.gradle
@@ -1,1 +1,2 @@
 rootProject.name='hail'
+include 'shadedazure'

--- a/hail/shadedazure/build.gradle
+++ b/hail/shadedazure/build.gradle
@@ -1,0 +1,57 @@
+buildscript {
+    repositories {
+        mavenCentral()
+    }
+}
+
+plugins {
+  id 'java'
+  id 'com.github.johnrengelman.shadow'
+}
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+    implementation group: 'com.azure', name: 'azure-storage-blob', version: '12.22.0'
+    implementation group: 'com.azure', name: 'azure-core-http-netty', version: '1.13.3'
+    implementation group: 'com.azure', name: 'azure-identity', version:'1.8.3'
+}
+
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+tasks.withType(ShadowJar) {
+    archiveBaseName = 'shadedazure'
+    zip64 true
+
+    // generate the jar then lightly prune the output of:
+    // jar tf shadedazure/build/libs/shadedazure-all.jar \
+    //     | awk -F'/' '{print "relocate '\''"$1"."$2"'\'', '\''is.hail.shadedazure."$1"."$2"'\''"}' \
+    //     | sort -u
+    relocate 'com.azure', 'is.hail.shadedazure.com.azure'
+    relocate 'com.ctc', 'is.hail.shadedazure.com.ctc'
+    relocate 'com.fasterxml', 'is.hail.shadedazure.com.fasterxml'
+    relocate 'com.microsoft', 'is.hail.shadedazure.com.microsoft'
+    relocate 'com.nimbusds', 'is.hail.shadedazure.com.nimbusds'
+    relocate 'com.sun', 'is.hail.shadedazure.com.sun'
+    relocate 'io.netty', 'is.hail.shadedazure.io.netty'
+    relocate 'is.hail', 'is.hail.shadedazure.is.hail'
+    relocate 'net.jcip', 'is.hail.shadedazure.net.jcip'
+    relocate 'net.minidev', 'is.hail.shadedazure.net.minidev'
+    relocate 'org.apache', 'is.hail.shadedazure.org.apache'
+    relocate 'org.codehaus', 'is.hail.shadedazure.org.codehaus'
+    relocate 'org.objectweb', 'is.hail.shadedazure.org.objectweb'
+    relocate 'org.reactivestreams', 'is.hail.shadedazure.org.reactivestreams'
+    relocate 'org.slf4j', 'is.hail.shadedazure.org.slf4j'
+    relocate 'reactor.adapter', 'is.hail.shadedazure.reactor.adapter'
+    relocate 'reactor.core', 'is.hail.shadedazure.reactor.core'
+    relocate 'reactor.netty', 'is.hail.shadedazure.reactor.netty'
+    relocate 'reactor.util', 'is.hail.shadedazure.reactor.util'
+
+    exclude 'META-INF/*.RSA'
+    exclude 'META-INF/*.SF'
+    exclude 'META-INF/*.DSA'
+}
+
+// you can make the jar from the parent directory with ./gradlew :shadedazure:shadowJar
+shadowJar {}

--- a/hail/src/main/resources/META-INF.services/com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF.services/com.azure.core.http.HttpClientProvider
@@ -1,1 +1,0 @@
-is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF.services/com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF.services/com.azure.core.http.HttpClientProvider
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF.services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF.services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
@@ -1,1 +1,0 @@
-is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF.services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF.services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF/services/com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF/services/com.azure.core.http.HttpClientProvider
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF/services/com.ctc.wstx.shaded.msv.relaxng_datatype.DatatypeLibraryFactory
+++ b/hail/src/main/resources/META-INF/services/com.ctc.wstx.shaded.msv.relaxng_datatype.DatatypeLibraryFactory
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.shaded.msv_core.datatype.xsd.ngimpl.DataTypeLibraryImpl

--- a/hail/src/main/resources/META-INF/services/com.ctc.wstx.stax.WstxInputFactory
+++ b/hail/src/main/resources/META-INF/services/com.ctc.wstx.stax.WstxInputFactory
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.stax.WstxInputFactory

--- a/hail/src/main/resources/META-INF/services/com.fasterxml.jackson.core.JsonFactory
+++ b/hail/src/main/resources/META-INF/services/com.fasterxml.jackson.core.JsonFactory
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.fasterxml.jackson.core.JsonFactory

--- a/hail/src/main/resources/META-INF/services/com.fasterxml.jackson.core.ObjectCodec
+++ b/hail/src/main/resources/META-INF/services/com.fasterxml.jackson.core.ObjectCodec
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.fasterxml.jackson.databind.ObjectMapper

--- a/hail/src/main/resources/META-INF/services/com.fasterxml.jackson.databind.Module
+++ b/hail/src/main/resources/META-INF/services/com.fasterxml.jackson.databind.Module
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.fasterxml.jackson.datatype.jsr310.JavaTimeModule

--- a/hail/src/main/resources/META-INF/services/io.micrometer.context.ContextAccessor
+++ b/hail/src/main/resources/META-INF/services/io.micrometer.context.ContextAccessor
@@ -1,0 +1,1 @@
+is.hail.shadedazure.reactor.netty.contextpropagation.ChannelContextAccessor

--- a/hail/src/main/resources/META-INF/services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF/services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
@@ -1,1 +1,0 @@
-is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF/services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
+++ b/hail/src/main/resources/META-INF/services/is.hail.shadedazure.com.azure.core.http.HttpClientProvider
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientProvider

--- a/hail/src/main/resources/META-INF/services/javax.xml.stream.XMLEventFactory
+++ b/hail/src/main/resources/META-INF/services/javax.xml.stream.XMLEventFactory
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.stax.WstxEventFactory

--- a/hail/src/main/resources/META-INF/services/javax.xml.stream.XMLInputFactory
+++ b/hail/src/main/resources/META-INF/services/javax.xml.stream.XMLInputFactory
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.stax.WstxInputFactory

--- a/hail/src/main/resources/META-INF/services/javax.xml.stream.XMLOutputFactory
+++ b/hail/src/main/resources/META-INF/services/javax.xml.stream.XMLOutputFactory
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.stax.WstxOutputFactory

--- a/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory
+++ b/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory
@@ -1,0 +1,3 @@
+is.hail.shadedazure.com.ctc.wstx.dtd.DTDSchemaFactory
+is.hail.shadedazure.com.ctc.wstx.msv.RelaxNGSchemaFactory
+is.hail.shadedazure.com.ctc.wstx.msv.W3CSchemaFactory

--- a/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory.dtd
+++ b/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory.dtd
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.dtd.DTDSchemaFactory

--- a/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory.relaxng
+++ b/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory.relaxng
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.msv.RelaxNGSchemaFactory

--- a/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory.w3c
+++ b/hail/src/main/resources/META-INF/services/org.codehaus.stax2.validation.XMLValidationSchemaFactory.w3c
@@ -1,0 +1,1 @@
+is.hail.shadedazure.com.ctc.wstx.msv.W3CSchemaFactory

--- a/hail/src/main/resources/META-INF/services/reactor.blockhound.integration.BlockHoundIntegration
+++ b/hail/src/main/resources/META-INF/services/reactor.blockhound.integration.BlockHoundIntegration
@@ -1,0 +1,1 @@
+is.hail.shadedazure.reactor.core.scheduler.ReactorBlockHoundIntegration

--- a/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
+++ b/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
@@ -1,10 +1,10 @@
 package is.hail.io.fs
 
-import com.azure.core.credential.TokenCredential
-import com.azure.identity.{ClientSecretCredential, ClientSecretCredentialBuilder, DefaultAzureCredential, DefaultAzureCredentialBuilder}
-import com.azure.storage.blob.models.{BlobProperties, BlobRange, ListBlobsOptions, BlobStorageException}
-import com.azure.storage.blob.specialized.BlockBlobClient
-import com.azure.storage.blob.{BlobClient, BlobContainerClient, BlobServiceClient, BlobServiceClientBuilder}
+import is.hail.shadedazure.com.azure.core.credential.TokenCredential
+import is.hail.shadedazure.com.azure.identity.{ClientSecretCredential, ClientSecretCredentialBuilder, DefaultAzureCredential, DefaultAzureCredentialBuilder}
+import is.hail.shadedazure.com.azure.storage.blob.models.{BlobProperties, BlobRange, ListBlobsOptions, BlobStorageException}
+import is.hail.shadedazure.com.azure.storage.blob.specialized.BlockBlobClient
+import is.hail.shadedazure.com.azure.storage.blob.{BlobClient, BlobContainerClient, BlobServiceClient, BlobServiceClientBuilder}
 import is.hail.services.retryTransientErrors
 import is.hail.io.fs.FSUtil.{containsWildcard, dropTrailingSlash}
 import org.apache.log4j.Logger

--- a/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
+++ b/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
@@ -5,6 +5,7 @@ import is.hail.shadedazure.com.azure.identity.{ClientSecretCredential, ClientSec
 import is.hail.shadedazure.com.azure.storage.blob.models.{BlobProperties, BlobRange, ListBlobsOptions, BlobStorageException}
 import is.hail.shadedazure.com.azure.storage.blob.specialized.BlockBlobClient
 import is.hail.shadedazure.com.azure.storage.blob.{BlobClient, BlobContainerClient, BlobServiceClient, BlobServiceClientBuilder}
+import is.hail.shadedazure.com.azure.core.http.netty.NettyAsyncHttpClientBuilder
 import is.hail.shadedazure.reactor.netty.http.client.HttpClient
 import is.hail.services.retryTransientErrors
 import is.hail.io.fs.FSUtil.{containsWildcard, dropTrailingSlash}

--- a/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
+++ b/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
@@ -135,7 +135,6 @@ class AzureBlobServiceClientCache(credential: TokenCredential) {
         val blobServiceClient = new BlobServiceClientBuilder()
           .credential(credential)
           .endpoint(s"https://$account.blob.core.windows.net")
-          .httpClient(new NettyAsyncHttpClientBuilder(HttpClient.create()).build())
           .buildClient()
         clients += ((account, container) -> blobServiceClient)
         blobServiceClient
@@ -145,7 +144,6 @@ class AzureBlobServiceClientCache(credential: TokenCredential) {
   def setPublicAccessServiceClient(account: String, container: String): Unit = {
     val blobServiceClient = new BlobServiceClientBuilder()
       .endpoint(s"https://$account.blob.core.windows.net")
-      .httpClient(new NettyAsyncHttpClientBuilder(HttpClient.create()).build())
       .buildClient()
     clients += ((account, container) -> blobServiceClient)
   }

--- a/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
+++ b/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
@@ -134,7 +134,7 @@ class AzureBlobServiceClientCache(credential: TokenCredential) {
         val blobServiceClient = new BlobServiceClientBuilder()
           .credential(credential)
           .endpoint(s"https://$account.blob.core.windows.net")
-          .httpClient(HttpClient.create())
+          .httpClient(new NettyAsyncHttpClientBuilder(HttpClient.create()).build())
           .buildClient()
         clients += ((account, container) -> blobServiceClient)
         blobServiceClient
@@ -144,7 +144,7 @@ class AzureBlobServiceClientCache(credential: TokenCredential) {
   def setPublicAccessServiceClient(account: String, container: String): Unit = {
     val blobServiceClient = new BlobServiceClientBuilder()
       .endpoint(s"https://$account.blob.core.windows.net")
-      .httpClient(HttpClient.create())
+      .httpClient(new NettyAsyncHttpClientBuilder(HttpClient.create()).build())
       .buildClient()
     clients += ((account, container) -> blobServiceClient)
   }

--- a/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
+++ b/hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
@@ -5,6 +5,7 @@ import is.hail.shadedazure.com.azure.identity.{ClientSecretCredential, ClientSec
 import is.hail.shadedazure.com.azure.storage.blob.models.{BlobProperties, BlobRange, ListBlobsOptions, BlobStorageException}
 import is.hail.shadedazure.com.azure.storage.blob.specialized.BlockBlobClient
 import is.hail.shadedazure.com.azure.storage.blob.{BlobClient, BlobContainerClient, BlobServiceClient, BlobServiceClientBuilder}
+import is.hail.shadedazure.reactor.netty.http.client.HttpClient
 import is.hail.services.retryTransientErrors
 import is.hail.io.fs.FSUtil.{containsWildcard, dropTrailingSlash}
 import org.apache.log4j.Logger
@@ -124,16 +125,16 @@ object AzureStorageFileStatus {
 }
 
 class AzureBlobServiceClientCache(credential: TokenCredential) {
-  private lazy val clientBuilder: BlobServiceClientBuilder = new BlobServiceClientBuilder()
-  private lazy val clients: mutable.Map[(String, String), BlobServiceClient] = mutable.Map()
+  private[this] lazy val clients = mutable.Map[(String, String), BlobServiceClient]()
 
   def getServiceClient(account: String, container: String): BlobServiceClient = {
     clients.get((account, container)) match {
       case Some(client) => client
       case None =>
-        val blobServiceClient = clientBuilder
+        val blobServiceClient = new BlobServiceClientBuilder()
           .credential(credential)
           .endpoint(s"https://$account.blob.core.windows.net")
+          .httpClient(HttpClient.create())
           .buildClient()
         clients += ((account, container) -> blobServiceClient)
         blobServiceClient
@@ -142,8 +143,9 @@ class AzureBlobServiceClientCache(credential: TokenCredential) {
 
   def setPublicAccessServiceClient(account: String, container: String): Unit = {
     val blobServiceClient = new BlobServiceClientBuilder()
-        .endpoint(s"https://$account.blob.core.windows.net")
-        .buildClient()
+      .endpoint(s"https://$account.blob.core.windows.net")
+      .httpClient(HttpClient.create())
+      .buildClient()
     clients += ((account, container) -> blobServiceClient)
   }
 }

--- a/hail/src/main/scala/is/hail/services/package.scala
+++ b/hail/src/main/scala/is/hail/services/package.scala
@@ -11,7 +11,7 @@ import org.apache.http.ConnectionClosedException
 import org.apache.http.conn.HttpHostConnectException
 import org.apache.log4j.{LogManager, Logger}
 
-import reactor.core.Exceptions.ReactiveException
+import is.hail.shadedazure.reactor.core.Exceptions.ReactiveException
 import scala.util.Random
 import java.io._
 import com.google.cloud.storage.StorageException
@@ -39,7 +39,7 @@ package object services {
     // An exception is a "retry once error" if a rare, known bug in a dependency or in a cloud
     // provider can manifest as this exception *and* that manifestation is indistinguishable from a
     // true error.
-    val e = reactor.core.Exceptions.unwrap(_e)
+    val e = is.hail.shadedazure.reactor.core.Exceptions.unwrap(_e)
     e match {
       case e: SocketException =>
         e.getMessage != null && e.getMessage.contains("Connection reset")
@@ -63,7 +63,7 @@ package object services {
     //
     // If the argument is a ReactiveException, it returns its cause. If the argument is not a
     // ReactiveException it returns the exception unmodified.
-    val e = reactor.core.Exceptions.unwrap(_e)
+    val e = is.hail.shadedazure.reactor.core.Exceptions.unwrap(_e)
     e match {
       case e: NoHttpResponseException =>
         true

